### PR TITLE
added input verification to duration parameter of piezo's buzz method

### PR DESCRIPTION
**Pull request by HU90m**
This fixes issues #147 and #122
I also increased the test coverage to 100%.

### CODE REVIEW

**Comment by kierdavis** [comment]
Thanks for looking at this! Unfortunately, I think I should have been clearer in #122. I was only intending for the PiezoInterface and its related backends to change, not the Piezo components, and without removing the `Pitch` type.

The aim is that the user passes a `Pitch` (i.e. a note or frequency) to `Piezo.buzz`. `Piezo.buzz` does some validation on these values, and if a note was passed it is converted to a frequency (the code for this already exists, we're just currently throwing the frequency away). This frequency is then passed to `PiezoInterface.buzz`. The end result is that note-to-frequency conversion only happens once (inside `Piezo.buzz` rather than in every backend), and `PiezoInterface` and backends only over have to deal with frequencies.

File: j5/components/piezo.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def interface_class() -> Type[PiezoInterface]:
 
     def buzz(self, duration: timedelta, pitch: Pitch) -> None:
         """Queue a note to be played."""
-        if type(pitch) is not int and type(pitch) is not Note:
+        if not (type(pitch) is int or type(pitch) is Note):
```
Comment (trickeydan) [inline j5/components/piezo.py]: The previous formatting of this is more pythonic.

The bot wants this method split up. I'd recommend creating a static `pitch_to_int` or `validate_pitch` method. Perhaps even both.

File: j5/components/piezo.py [hunk 2]
```diff
@@ -67,8 +67,15 @@ def interface_class() -> Type[PiezoInterface]:
 
     def buzz(self, duration: timedelta, pitch: Pitch) -> None:
         """Queue a note to be played."""
-        if not (type(pitch) is int or type(pitch) is Note):
+        if type(pitch) is int:
+            pass
+        elif type(pitch) is Note:
```
Comment (trickeydan) [inline j5/components/piezo.py]: See other comment about complexity

File: j5/components/piezo.py (hunk 1 above)
Comment (HU90m) [reply j5/components/piezo.py]: Ah didn't think of that, thank you!

Comment (HU90m) [reply j5/components/piezo.py]: Sorry, I didn't read this properly first time round. I have now done this.

File: j5/components/piezo.py [hunk 3]
```diff
@@ -67,14 +67,21 @@ def interface_class() -> Type[PiezoInterface]:
...
             raise ValueError("Pitch must be greater than zero")
+        elif duration < timedelta(seconds=0):
+            raise ValueError("Duration must be greater than zero")
         else:
             self._backend.buzz(self._board, self._identifier, duration, pitch)
+
+    @staticmethod
+    def verify_pitch(pitch: Pitch) -> bool:
+        """Verifies a pitch is either a int or Note."""
+        pitch_is_not_an_int = type(pitch) is not int
```
Comment (HU90m) [inline j5/components/piezo.py]: Are the following variables required/necessary?
```
        pitch_is_not_an_int = type(pitch) is not int
        pitch_is_not_a_Note = type(pitch) is not Note
```

File: j5/components/piezo.py [hunk 4]
```diff
@@ -67,14 +67,21 @@ def interface_class() -> Type[PiezoInterface]:
...
-        if frequency < 0:
+        if pitch < 0:
             raise ValueError("Pitch must be greater than zero")
+        elif duration < timedelta(seconds=0):
+            raise ValueError("Duration must be greater than zero")
         else:
             self._backend.buzz(self._board, self._identifier, duration, pitch)
+
+    @staticmethod
+    def verify_pitch(pitch: Pitch) -> bool:
```
Comment (trickeydan) [inline j5/components/piezo.py]:
```suggestion
def verify_pitch(pitch: Pitch) -> None:
        """Verify that a pitch is valid."""
        # Verify that the type is correct.
        pitch_is_int = type(pitch) is  int
        pitch_is_note = type(pitch) is Note
        if not (pitch_is_int or pitch_is_note):
            raise TypeError("Pitch must be int or Note")
       
       # Verify the length of the pitch is non-zero
       if pitch < 0:
           raise ValueError("Frequency must be greater than zero")
           
@staticmethod
def verify_duration(duration: timedelta):
    """Verify that a duration is valid."""
    if not instance(duration, timedelta):
        raise TypeError("Duration must be of type datetime.timedelta")
    if duration < timedelta(seconds=0):
        raise ValueError("Duration must be greater than or equal to zero.")
```

Comment (trickeydan) [reply j5/components/piezo.py]: You will need to adjust your other code to use these:
e.g
```
verify_pitch(pitch)
verify_duration(duration)
self._backend.buzz(self._board, self._identifier, duration, pitch)
```

**Pull request merged**